src/syntax/sugar.rkt

```diff
@@ -179,65 +173,45 @@
-                                          (representation-name repr)))
-            (values (list cast expr) repr)])]
+        [(? variable?) (values expr (context-lookup ctx expr))]
         [(? number?)
          (define prec (representation-name (context-repr ctx)))
          (define num
            (match expr
              [(or +inf.0 -inf.0 +nan.0) expr]
              [(? exact?) expr]
              [_ (inexact->exact expr)]))
-         (values (literal num prec) (context-repr ctx))]
-        [(? boolean?) (values expr (get-representation 'bool))])))
+         (values (literal num prec) (context-repr ctx))])))
   expr*)
 
-;; Translates an ImplProg to an FPCore.
-(define (prog->fpcore prog repr)
-  (let loop ([expr prog])
-    (match expr
-      [`(if ,cond ,ift ,iff) `(if ,(loop cond) ,(loop ift) ,(loop iff))]
-      [`(,(? cast-impl? impl) ,body)
-       (match-define (list irepr) (impl-info impl 'itype))
-       (define body* (prog->fpcore body irepr))
-       (cond
-         [(list? body*) `(cast (! :precision ,(representation-name irepr) ,body*))]
-         [else body*])] ; constants and variables should not have casts
-      [`(,impl) (impl->operator impl)]
-      [`(,impl ,args ...)
-       (define op (impl->operator impl))
-       (define args* (map prog->fpcore args (impl-info impl 'itype)))
-       (match (cons op args*)
-         [`(neg ,arg) `(- ,arg)]
-         [expr expr])]
-      [(? variable?) expr]
-      [(? literal?)
-       (match (literal-value expr)
-         [-inf.0 '(- INFINITY)]
-         [+inf.0 'INFINITY]
-         [+nan.0 'NAN]
-         [v
-          (if (set-member? '(binary64 binary32) (literal-precision expr)) (exact->inexact v) v)])])))
-
-;; Translates an ImplProg to a Spec.
-(define (prog->spec expr)
+;; Translates from LImpl to an FPCore.
+(define (prog->fpcore expr)
   (match expr
-    [`(if ,cond ,ift ,iff) `(if ,(prog->spec cond) ,(prog->spec ift) ,(prog->spec iff))]
-    [`(,(? cast-impl? impl) ,body) `(,impl ,(prog->spec body))]
+    [`(if ,cond ,ift ,iff) `(if ,(prog->fpcore cond) ,(prog->fpcore ift) ,(prog->fpcore iff))]
+    [`(,(? cast-impl? impl) ,body)
+     (define prec (representation-name (impl-info impl 'otype)))
+     `(! :precision ,prec (cast ,(prog->fpcore body)))]
+    [`(,impl) (impl->operator impl)]
     [`(,impl ,args ...)
      (define op (impl->operator impl))
-     (define args* (map prog->spec args))
-     `(,op ,@args*)]
+     (define args* (map prog->fpcore args))
+     (match (cons op args*)
+       [`(neg ,arg) `(- ,arg)]
+       [expr expr])]
     [(? variable?) expr]
-    [(? literal?) (literal-value expr)]))
+    [(? literal?)
+     (match (literal-value expr)
+       [-inf.0 '(- INFINITY)]
+       [+inf.0 'INFINITY]
+       [+nan.0 'NAN]
+       [v (if (set-member? '(binary64 binary32) (literal-precision expr)) (exact->inexact v) v)])]))
+
+;; Translates an LImpl to a LSpec.
+(define (prog->spec expr)
+  (expand-accelerators
```

Comment:
Ugly we need to call `expand-accelerators` here instead of not having operators at all and using the spec here, but yeah that's fine for now.

Reply:
Yep. Hopefully that'll be fixed soon!